### ISSUE TITLE
Question: dynamic header injection

### ISSUE THREAD

**Issue by danielloader** [FKLC/AnyAPI#2]
Hey thanks for the library, looks handy but want to clarify a point before I try to implement something.

Right now I'm using nginx as a non secure to secure API bridge, that is to say, requests originating from localhost can hit nginx and a python script will go and get an oauth2 bearer token and inject it into the header so what is an unauthenticated API call can be proxied to a secure endpoint.

Is that possible using this library? I'd like to take nginx out of the equation if I'm already using python to monitor the bearer token and expiry and subsequent refresh and injection of the header into the nginx.conf file.

**Comment by FKLC** [comment]
I can't say that I really understand your question because of my lack knowledge of nginx but if you are asking for how to update/set headers dynamically you should see this wiki page https://github.com/FKLC/AnyAPI/wiki/3.-Filters

**Comment by danielloader** [comment]
Well to outline what happens, in one python script I request a bearer token from an external API, I get this token back and set it as a header on the nginx config - this config is then hot reloaded and requests then use this new Authorization: Bearer $TOKEN header to proxy requests back to the API in question.

Nginx does safe reloading where current http transactions are finished before it swaps to a new configuration, so anything currently outstanding or currently in progress would continue using the old bearer and configuration until completion, while new requests would come in and be proxied with a new bearer auth header.

So basically I'm just wondering how graceful this would be if I change the header programmatically (I know in your example you add a date header dynamically). Basically can't afford for proxied requests to be killed at the time of bearer token change (prior to the natural expiry).

If you need any other details to clarify any points I'll just do a flowchart or something.  

Thanks a lot!

**Comment by FKLC** [comment]
1. What do mean by "kill" do you want to abort an ongoing request if so according to my search this is hard to implement.

2. If you understand how to set headers programmatically so what is the problem why you can't set `keyword_arguments['headers']['Authorization'] = 'Bearer your_token'`

I guess the problem is that I still don't understand your question I think a flowchart would be good

**Comment by danielloader** [comment]
I think point 2 is enough for me to give an implementation a go instead and continue this if need be, as far as I can tell this will in fact do everything I need it to do! 

Is there any method of authenticating requests before they're proxied on? Or will I still need nginx in front just handling authentication instead of header changes?

Thanks for replying though!

**Comment by FKLC** [comment]
Since AnyAPI built on [requests](https://2.python-requests.org//en/master/) you can use all of the auth plugins out there for [requests](https://2.python-requests.org//en/master/).

If your API is using OAuth2 you can use [requests-oauthlib](https://2.python-requests.org//en/master/user/authentication/#oauth-2-and-openid-connect-authentication) or you can [implement your own auth](https://2.python-requests.org//en/master/user/authentication/#oauth-2-and-openid-connect-authentication)

Usage:
```py
my_api = AnyAPI(base_url, auth=your_auth_handler)
```

**Comment by danielloader** [comment]
Perfect thanks for this information, I'll give it a go!
